### PR TITLE
Fix broken webhooks

### LINKED ISSUE #29689: WechatWork webhook can NOT use after apply PR:[REFACTOR] store webhook event in database
### Description

WechatWork webhook can NOT use after apply PR:[REFACTOR] store webhook event in database

https://github.com/go-gitea/gitea/pull/29145

wechatwork webhook robot response : "errcode":40008,"errmsg":"invalid message type".

then, I try to revert this pr, the wechatwork webhook work correctly.

### Gitea Version

main branch

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

8021d316d5a45aeabb51085302cd3d42

e3379a68b06ba9210b782ccf86ff2719

### Git Version

main branch

### Operating System

windows

### How are you running Gitea?

windows exe

### Database

MSSQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #29689 
This PR makes the default http method of webhook as POST instead of GET

### CODE REVIEW

**Comment by wxiaoguang** [comment]
How does this change fix the bug? `newDefaultRequest` uses POST by default by a fallthrough.

**Comment by wxiaoguang** [comment]
And I can see JS error on the "add webhook" page

```
WebHookEditor.js:26 Uncaught TypeError: Cannot read properties of null (reading 'value')
    at updateContentType (WebHookEditor.js:26:58)
    at initCompWebHookEditor (WebHookEditor.js:29:3)
    at HTMLDocument.<anonymous> (index.js:109:24)

  const updateContentType = function () {
    const visible = document.getElementById('http_method').value === 'POST';  // <--------- HERE
    toggleElem(document.getElementById('content_type').parentNode.parentNode, visible);
  };
```

Maybe something is caused by the refactoring, while something is already broken before the refactoring.

**Comment by lunny** [comment]
> How does this change fix the bug? `newDefaultRequest` uses POST by default by a fallthrough.

I have tested in production env, it fixed. I haven't looked into the logic.

**Comment by wxiaoguang** [comment]
I think the real bug is here:

image

**Comment by wxiaoguang** [comment]
> I have tested in production env, it fixed. I haven't looked into the logic.

I guess you only tested WeCom(WechatWork)

But the underlying bug affects almost all webhooks.

**Comment by lunny** [comment]
> > I have tested in production env, it fixed. I haven't looked into the logic.
> 
> I guess you only tested WeCom(WechatWork)
> 
> But the underlying bug affects almost all webhooks.

The `w.HTTPMethod` should be `POST` but now it's GET and the database field `http_method` is empty. Some webhooks types accept `GET` like feishu/lark, so they haven't been affected.

**Comment by wxiaoguang** [comment]
> > > I have tested in production env, it fixed. I haven't looked into the logic.
> > 
> > 
> > I guess you only tested WeCom(WechatWork)
> > But the underlying bug affects almost all webhooks.
> 
> The `w.HTTPMethod` should be `POST` but now it's GET and the database field `http_method` is empty. Some webhooks types accept `GET` like feishu/lark, so they haven't been affected.

I think the logic should be checked carefully. Do all these webhooks accept "GET"? Should it default the empty HTTP method to POST (just like newDefaultRequest)?

image

**Comment by lunny** [comment]
> > > > I have tested in production env, it fixed. I haven't looked into the logic.
> > > 
> > > 
> > > I guess you only tested WeCom(WechatWork)
> > > But the underlying bug affects almost all webhooks.
> > 
> > 
> > The `w.HTTPMethod` should be `POST` but now it's GET and the database field `http_method` is empty. Some webhooks types accept `GET` like feishu/lark, so they haven't been affected.
> 
> I think the logic should be checked carefully. Do all these webhooks accept "GET"? Should it default the empty HTTP method to POST (just like newDefaultRequest)?
> 
> image

Maybe we need to fill HTTPMethod clearly for every webhookParams. At present, `Gitea`, `matrix` filled the HTTPMethod when creating.

图片

**Comment by wxiaoguang** [comment]
But I do not see why it should use "GET" method for a "JSON body"

**Comment by silverwind** [comment]
Please don't use `GET` with body. Either move params to query or make it a `POST`.

**Comment by lunny** [comment]
> Please don't use `GET` with body, unless the other end requires it. Either move params to query or make it a `POST`.

I believe #29145 accidently make the default methods as `GET`, we are talking about how to fix it.

**Comment by wxiaoguang** [comment]
> > But I do not see why it should use "GET" method for a "JSON body"
> 
> I just pointed out that feishu/lark hasn't been affected by the `POST` -> `GET` change because they support `GET`. But I don't mean it should be `GET`.

I was talking about https://github.com/go-gitea/gitea/pull/29690#issuecomment-1986850374 : after the feishu/lark comment, it is for the `Maybe we need to fill HTTPMethod clearly for every webhookParams. At present, Gitea, matrix filled the HTTPMethod when creating`, my suggestion is `use "POST" for "JSON body" in the "newJSONRequest"` (if it is the old behavior)

**Comment by lunny** [comment]
> > > But I do not see why it should use "GET" method for a "JSON body"
> > 
> > 
> > I just pointed out that feishu/lark hasn't been affected by the `POST` -> `GET` change because they support `GET`. But I don't mean it should be `GET`.
> 
> I was talking about [#29690 (comment)](https://github.com/go-gitea/gitea/pull/29690#issuecomment-1986850374) : after the feishu/lark comment, it is for the `Maybe we need to fill HTTPMethod clearly for every webhookParams. At present, Gitea, matrix filled the HTTPMethod when creating`, my suggestion is `use "POST" for "JSON body" in the "newJSONRequest"` (if it is the old behavior)

6052ecf14fb2258e267ca87a31a6e58ae7571158

File: routers/web/repo/setting/webhook.go [hunk 1]
```diff
@@ -551,6 +551,7 @@ func wechatworkHookParams(ctx *context.Context) webhookParams {
 		Type:        webhook_module.WECHATWORK,
 		URL:         form.PayloadURL,
 		ContentType: webhook.ContentTypeJSON,
+		HTTPMethod:  http.MethodPost,
```
Comment (wxiaoguang) [inline routers/web/repo/setting/webhook.go]: It should be reverted.

Comment (lunny) [reply routers/web/repo/setting/webhook.go]: Done.

**Review by silverwind — APPROVED** [review]

**Comment by jlw4049** [comment]
Can confirm that latest nightly broke discord webhooks as well

**Review by jolheiser — APPROVED** [review]

**Pull request merged**